halo2_proofs/src/plonk/circuit.rs

```diff
@@ -705,7 +808,44 @@ pub enum Expression<F> {
     Scaled(Box<Expression<F>>, F),
 }
 
+
+
 impl<F: Field> Expression<F> {
+
+    /// Make side effects
+    pub fn load_to_virtual_cells(&mut self, cells: &mut VirtualCells<'_, F>) {
+        match self {
+            Expression::Constant(_) => (),
+            Expression::Selector(_) => (),
+            Expression::Fixed(query) => {
+                let col = Column{ index: query.column_index, column_type: Fixed};
+                cells.queried_cells.push((col, query.rotation).into());
+                query.index = Some(cells.meta.query_fixed_index(col, query.rotation));
+            },
+            Expression::Advice(query) => {
+                let col = Column{ index: query.column_index, column_type: Advice{phase: query.phase}};
+                cells.queried_cells.push((col, query.rotation).into());
+                query.index = Some(cells.meta.query_advice_index(col, query.rotation));
+            }
+            Expression::Instance(query) => {
+                let col = Column{ index: query.column_index, column_type: Instance};
+                cells.queried_cells.push((col, query.rotation).into());
+                query.index = Some(cells.meta.query_instance_index(col, query.rotation));
+            },
```

Comment:
Is it possible to simply reuse `query_fixed`, `query_advice` and `query_instance` here? Because now the query index is set in two different places using two versions of the code which, admittedly, is currently very short, but still I think it would be great to only do it in one place to make sure the same logic is always done no matter how the cell is queried. I would think something like `self = cells.query_advice(...)` could work.

Reply:
Or alternatively, modify `query_fixed` etc... to not set the query index anymore either. That's probably better because then the Expressions always have the same state no matter what API the user used when they get passed into halo2, so there's no 2 possible scenario's anymore. This also means we don't _really_ need additional tests anymore (except just using the new API once somewhere) as all methods of querying are basically the same.

Reply:
So I'll leave it the same in this recursive method and change the cells.query_X().
Yea you're right, with this PR merged in there will be incompatibility where ConstraintSystem.advice_queries.push((column, at)) will be called twice.

Reply:
So it's better to change the old API so that we only push the queries at the end of create_gate

Reply:
Yeah I think so